tests/0114-sticky_partitioning.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+ * librdkafka - Apache Kafka C library
+ *
+ * Copyright (c) 2020, Magnus Edenhill
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * Test sticky.partitioning.linger.ms producer property.
+ *
+ */
+
+#include <iostream>
+#include <fstream>
+#include <iterator>
+#include <string>
+#include "testcpp.h"
+#include "test.h"
+
+/**
+ * @brief Specify sticky.partitioning.linger.ms and check consumed
+ * messages to verify it takes effect.
+ */
+static void do_test_sticky_partitioning (int sticky_delay) {
+  
+  std::string topic = Test::mk_topic_name(__FILE__, 1);
+  Test::create_topic(NULL, topic.c_str(), 3, 1);
+
+  RdKafka::Conf *conf;
+  Test::conf_init(&conf, NULL, 0);
+
+  Test::conf_set(conf, "sticky.partitioning.linger.ms", tostr() << sticky_delay);
+
+  std::string errstr;
+  RdKafka::Producer *p = RdKafka::Producer::create(conf, errstr);
+  if (!p)
+          Test::Fail("Failed to create Producer: " + errstr);
+
+  RdKafka::Consumer *c = RdKafka::Consumer::create(conf, errstr);
+  if (!c)
+          Test::Fail("Failed to create Consumer: " + errstr);
+
+  RdKafka::Topic *t = RdKafka::Topic::create(c, topic, NULL, errstr);
+  if (!t)
+          Test::Fail("Failed to create Topic: " + errstr);
+
+  c->start(t, 0, RdKafka::Topic::OFFSET_BEGINNING);
+  c->start(t, 1, RdKafka::Topic::OFFSET_BEGINNING);
+  c->start(t, 2, RdKafka::Topic::OFFSET_BEGINNING);
+
+  const int msgcnt = 50000;
+  const int msgsize = 1000;
+
+  /* Produce messages */
+  char val[msgsize];
+  memset(val, 'a', msgsize);
+
+  for (int i = 0 ; i < msgcnt ; i++) {
```

Comment:
Without any pacing in this loop I don't see how we can guarantee that the sticky partition interval is hit during the producing of these messages, which will possibly make the test unstable and dependent on the runtime environment (CI being super slow and local desktop being super fast).

Instead I recommend for a fixed rate, e.g., 100 msgs/second, and to use sticky lingering values as multiples of that rate, allowing us to calculate the expected number of different partitions hit.